Add more filters for CodeExplorer import command

IvenBach opened the pull request:
Close #5395 
Add additional filters. Cache filter index for reuse next time dialog is shown.

IvenBach committed:
Add more filters for CodeExplorer import command

Add additional filters. Cache filter index for reuse next time
dialog is shown.

bclothier commented on Rubberduck.Core/UI/CodeExplorer/Commands/ImportCommand.cs:
```diff
@@ -562,6 +571,18 @@ protected override void OnExecute(object parameter)
+            ImportableExtensions.Where(ext => !ext.EndsWith("doccls"))
+                                .Select(ext => $"*{ext}"));
+
+        private string StandardModuleExtension => "*" + ImportableExtensions.Where(ext => ext.EndsWith("bas")).First();
```
My concern with using `First()` is that it becomes a runtime error that will blow upmif the assumption about file extensions being unique and present is violated.

I probably would use ‘FirstOrDefault()` and if `null`, do not concatenate into the filter. Then I would add an unit test to assert that the contents are unique and has valid extensions, thus preventing a time bomb.

IvenBach committed:
Use FirstOrDefault to eliminate runtime error

Add member to create individual filters and member to concatenate
individual filter for dialogs filter.

IvenBach committed:
Change members to be local functions

Converted members to local functions since both members are used in only one
location. They can only be used within the containing member and simplifies
the member list on the type.

IvenBach committed:
Add caching and persisting for filter index

Caches filter index last chosen for next import. Filter index is saved
to config file to continue using last chosen index went starting a new
instance of RD.

IvenBach committed:
Test Import command for filter index change

When filter index matches cached/config index no save occurs.
When filter index differs from cached/config index save occurs.

IvenBach committed:
Tests for behavior when import canceled

Cancelling an import does not change filter index. Saving does not occur.

IvenBach committed:
Fix failing unit tests

Restore inital null check and model mockProjectSettingsProvider after it.
Constructor inject dependencies.

IvenBach committed:
Merge remote-tracking branch 'upstream/next' into Issue5395_Filters_for_synchronizing_document_types

retailcoder commented:
Marking this PR as blocked for now, since it introduces new resource keys.

IvenBach commented:
Does the same go for the other resource editing PRs?

retailcoder commented:
@IvenBach I haven't looked at the conflicts, do they look salvageable? If they become mergeable after the release, I'll merge them, but the "resource freeze" window is closing soon, and then every new resource key will probably conflict again.

IvenBach commented:
Most of them were minor at worst. Will review each one to update then push for review.

retailcoder commented:
You'll probably want to resolve the conflicts on #5749 first, have it merged (I'd like to merge it sooner than later!), and then sync [next] with your dev branch here and address *these* conflicts.

retailcoder commented:
Closing as stale for now
